kv macro support

yoshuawuyts opened the pull request:
Adds key-value support to the log macros. Supersedes #346. Refs #343 #328. Thanks!

## Syntax
```rust
femme::start(log::LevelFilter::Info).unwrap();
info!("hello");
info!("hello",);
info!("hello {}", "cats");
info!("hello {}", "cats",);
info!("hello {}", "cats", {
    cat_1: "chashu",
    cat_2: "nori",
});
```


## Tasks
- [ ]  write tests

KodrAus commented on src/lib.rs:
```diff
@@ -1384,11 +1384,64 @@ pub fn logger() -> &'static Log {
     args: fmt::Arguments,
     level: Level,
     &(target, module_path, file, line): &(&str, &'static str, &'static str, u32),
+    kvs: Option<&[(&str, &str)]>,
```
We should use `&ToValue` here so that the structure of captured values can be retained. I sketched out the change in https://github.com/yoshuawuyts/kv-log-macro/pull/1

yoshuawuyts replied:
Ohhhh, yesss! -- I wish I'd known about this sooner, haha.

KodrAus commented:
Ah, we should also be careful not to regress the `info!("Hello, {value}", value = 42);` case, just based on [`kv-log-macro`](https://github.com/yoshuawuyts/kv-log-macro) it looks like our changes to the macro syntax may not support the `ident = expr` syntax?

It does also present a bit of a syntactic oddity, where values interpolated into the message would use the `ident = expr` syntax from `format_args`, but the captured key-value pairs use the `ident: expr` syntax like structs:

```rust
info!("hello {value}", value = "cats", {
    cat_1: "chashu",
    cat_2: "nori",
});
```

I don't think this is necessarily a problem though, but thought it was worth pointing out.

yoshuawuyts commented:
@KodrAus in general it seems we don't have a lot of tests right now -- I was wondering how we might want to go about it?

KodrAus commented:
> we don't have a lot of tests right now

Yeh, we don't really for the macro syntax, do we. Maybe we could add an integration test file that can do something like this:

```rust
struct AssertRecord;

impl Log for AssertRecord {
    fn record(&self, record: &Record) {
        ASSERT.with(|a| a(record));
    }
}

#[test]
fn level_info() {
    // Sets the assert function in thread-local storage
    // then calls the log statement
    test_log!(
        info!("Hello, {value}!", value = "world"),
        |record| {
            assert_eq!("Hello, world!", record.args().to_string();
            assert_eq!(Level::Info, record.level());
        }
    );
}
```

So that we can build up a suite for the current syntax.

But I'm happy to sketch that out independently if you like? For now I think if we can try find a way to keep `ident = expr` interpolated values working then we _should_ be good!

yoshuawuyts commented:
@KodrAus ohhh, yeah looks like you were onto something there -- I found a bug in my impl which I've fixed, but more importantly the `name = val` interpolated args aren't indeed no longer working. I've added some basic tests for the macros to prevent this from regressing.

Though I'm not entirely sure how to fix it, so I'm all ears if you have suggestions!

KodrAus commented:
Hi @yoshuawuyts!

I'm just circling back to this, I'll spend some time playing with the macros too and see where I end up. We might end up having to tweak the syntax a little (which we could do to consider working more nicely with other extensions like #357).

yoshuawuyts commented:
@KodrAus did you have any luck getting this to work? If not, perhaps we could ask @dtolnay for help?

KodrAus commented:
Hi @yoshuawuyts! :wave:

Sorry I’ve been really scattered over the last few months so haven’t been able to make a lot of progress here. However I’m feeling less concerned about regressing this specific edge-case, and I also think [implicit inline format args](https://github.com/rust-lang/rfcs/pull/2795) makes the difference between our binding syntax for format args vs kv pairs much less significant.

So all that’s to say I’d be on board with pushing forward with this!

yoshuawuyts commented:
@KodrAus yay, that's excellent news! I'm very excited for this!

I just pushed a patch to remove the failing tests, meaning this patch should now be ready to be merged! It seems CI is currently not setup, so it can't verify this patch. But hopefully this is good!

KodrAus commented:
> It seems CI is currently not setup

Oh dear... Lemme just fix that real quick.

KodrAus commented:
@yoshuawuyts would you like to squash and rebase your commits on the current `master`? I'm thinking the reason actions aren't running here is because we're missing the `.github/workflows` directory on your branch.

yoshuawuyts commented:
@KodrAus rebased and force-pushed. Hoping this works! :crossed_fingers:

yoshuawuyts commented:
Heh, for odd reasons CI is no longer reporting on this branch. My guess is because the repo was migrated it caused some bug in GitHub's UI.

CI link is here: https://github.com/yoshuawuyts/log/runs/295531981 -- it seems `cfg_if` is failing to build?

KodrAus commented:
Ah it looks like `cfg-if` has removed their explicit MSRV, so I'm trying to get everything back into a good state in #362.

yoshuawuyts commented:
Not behind a computer right now but we should try running CI again. Looks like #362 got merged two weeks ago.

yoshuawuyts commented:
@kodraus Wooot! So happy this got merged! Thank *you*!

Do you have any idea when this could make it to a release on crates.io? I'm super stoked to start trying it out! ✨

KodrAus commented:
@yoshuawuyts Yeh this is an exciting step forward! I'll put together a release PR today and cc you on it. I might get a few other peeps from Libs to take a look at the diff too just to make sure we haven't missed anything.
